distributed/diagnostics/progress.py

```diff
@@ -291,3 +293,94 @@ def transition(self, key, start, finish, *args, **kwargs):
     def restart(self, scheduler):
         self.all.clear()
         self.state.clear()
+
+
+class GroupTiming(SchedulerPlugin):
+    """Keep track of high-level timing information for task group progress"""
```

Comment:
Thoughts on putting some of this information directly on the TaskGroup?

Maybe this gets complicated because we clean up TaskGroups?

Reply:
I waffled on whether to put this in a plugin or directly on the task group, but went with this for a few reasons:
1. As you say, we clean up task groups, so that makes it somewhat trickier to get a longitudinal view on them.
2. I still view the specific data structures I'm building here to be pretty provisional, and thought that a plugin like this allowed for more isolation and iteration while hammering that out. Changes here are much easier to revert without API changes.
3. I'm still unsure about the performance characteristics of some of the things we are thinking of tracking, and the plugin approach is easier to only register on-demand.
4. Since this data isn't currently used for any scheduling logic, it didn't *need* to be on the task group (though in the future that could change!)

All that said, I don't have strong priors here, and I'm happy to be talked into moving this data onto the scheduler or task group.

Reply:
My guess is that if we can make them efficient that we'll want to eventually move them to TaskGroups, but leaving this until things are more stable sounds good to me.